### PR TITLE
Tobin inuse cleanup

### PR DESCRIPTION
**Pull request by tobine**
This adds the framework to fix the remaining in-use and cmd buffer invalid holes. Not all of the checks are added yet. Initially this adds cmd buffer binding for children for framebuffer (renderPass, imageViews & images) and imageView in general as well as in-use checks for imageView. Added tests for those two cases.

Will add more checks and test for remaining holes but wanted to submit this as initial PR since it's already getting large.

### CODE REVIEW

File: layers/core_validation.cpp [hunk 1]
```diff
@@ -6115,11 +6115,21 @@ DestroyFramebuffer(VkDevice device, VkFramebuffer framebuffer, const VkAllocatio
...
-    dev_data->device_dispatch_table->DestroyRenderPass(device, renderPass, pAllocator);
+    auto rp_state = getRenderPass(dev_data, renderPass);
+    if (rp_state) {
+        VK_OBJECT obj_struct = {reinterpret_cast<uint64_t &>(renderPass), VK_DEBUG_REPORT_OBJECT_TYPE_RENDER_PASS_EXT};
+        skip |= ValidateObjectNotInUse(dev_data, rp_state, obj_struct);
+        // Any bound cmd buffers are now invalid
+        invalidateCommandBuffers(rp_state->cb_bindings, obj_struct);
+    }
+    if (!skip) {
+        // TODO: leaking all the guts of the renderpass node here!
```
Comment (chrisforbes) [inline layers/core_validation.cpp]: Is this still true?

File: tests/vkrenderframework.cpp [hunk 2]
```diff
@@ -269,7 +269,7 @@ void VkRenderFramework::InitRenderTarget(uint32_t targets, VkImageView *dsBindin
     att.storeOp = VK_ATTACHMENT_STORE_OP_STORE;
     att.stencilLoadOp = VK_ATTACHMENT_LOAD_OP_DONT_CARE;
     att.stencilStoreOp = VK_ATTACHMENT_STORE_OP_DONT_CARE;
-    att.initialLayout = VK_IMAGE_LAYOUT_COLOR_ATTACHMENT_OPTIMAL;
+    att.initialLayout = VK_IMAGE_LAYOUT_UNDEFINED;
```
Comment (chrisforbes) [inline tests/vkrenderframework.cpp]: :+1:

File: tests/layer_validation_tests.cpp [hunk 3]
```diff
@@ -6869,6 +6869,96 @@ TEST_F(VkLayerTest, InvalidCmdBufferFramebufferImageDestroyed) {
     vkFreeMemory(m_device->device(), image_memory, nullptr);
 }
 
+TEST_F(VkLayerTest, FramebufferImageInUseDestroyedSignaled) {
+    TEST_DESCRIPTION("Delete in-use image that's child of framebuffer.");
+    VkFormatProperties format_properties;
+    VkResult err = VK_SUCCESS;
+    vkGetPhysicalDeviceFormatProperties(gpu(), VK_FORMAT_B8G8R8A8_UNORM, &format_properties);
+    if (!(format_properties.optimalTilingFeatures & VK_FORMAT_FEATURE_SAMPLED_IMAGE_BIT)) {
```
Comment (chrisforbes) [inline tests/layer_validation_tests.cpp]: Don't need this. You never actually sample from the image. [The format is guaranteed supported as a color attachment]

File: tests/layer_validation_tests.cpp [hunk 4]
```diff
@@ -6869,6 +6869,96 @@ TEST_F(VkLayerTest, InvalidCmdBufferFramebufferImageDestroyed) {
...
+    image_ci.imageType = VK_IMAGE_TYPE_2D;
+    image_ci.format = VK_FORMAT_B8G8R8A8_UNORM;
+    image_ci.extent.width = 256;
+    image_ci.extent.height = 256;
+    image_ci.extent.depth = 1;
+    image_ci.mipLevels = 1;
+    image_ci.arrayLayers = 1;
+    image_ci.samples = VK_SAMPLE_COUNT_1_BIT;
+    image_ci.tiling = VK_IMAGE_TILING_OPTIMAL;
+    image_ci.usage = VK_IMAGE_USAGE_COLOR_ATTACHMENT_BIT | VK_IMAGE_USAGE_SAMPLED_BIT;
```
Comment (chrisforbes) [inline tests/layer_validation_tests.cpp]: Just COLOR_ATTACHMENT_BIT

File: tests/layer_validation_tests.cpp [hunk 5]
```diff
@@ -6869,6 +6869,96 @@ TEST_F(VkLayerTest, InvalidCmdBufferFramebufferImageDestroyed) {
...
+    VkResult err = VK_SUCCESS;
+    vkGetPhysicalDeviceFormatProperties(gpu(), VK_FORMAT_B8G8R8A8_UNORM, &format_properties);
+    if (!(format_properties.optimalTilingFeatures & VK_FORMAT_FEATURE_SAMPLED_IMAGE_BIT)) {
+        return;
+    }
+
+    ASSERT_NO_FATAL_FAILURE(InitState());
+    ASSERT_NO_FATAL_FAILURE(InitRenderTarget());
+
+    VkImageCreateInfo image_ci = {};
```
Comment (chrisforbes) [inline tests/layer_validation_tests.cpp]: Is unfortunate that you have to abandon /all/ the helpers for this to test DestroyImage interactions...

File: layers/core_validation.cpp (hunk 1 above)
Comment (tobine) [inline layers/core_validation.cpp]: yes, unfortunately. We "new" all the stuff in the renderPass createInfo at create time and never delete.

File: layers/core_validation_types.h [hunk 6]
```diff
@@ -186,6 +186,18 @@ class IMAGE_NODE : public BASE_NODE {
     };
 };
 
+class IMAGE_VIEW_STATE : public BASE_NODE {
```
Comment (tobine) [inline layers/core_validation_types.h]: Changing convention. No more NODES going fwd and will bulk-rename others at some point.

File: layers/core_validation_types.h [hunk 7]
```diff
@@ -160,6 +160,18 @@ class BUFFER_NODE : public BASE_NODE {
           createInfo(rh_obj.createInfo){};
 };
 
+class BUFFER_VIEW_STATE : public BASE_NODE {
+  public:
+    using BASE_NODE::in_use;
```
Comment (tobine) [inline layers/core_validation_types.h]: Yep, removed.

File: layers/core_validation.cpp [hunk 8]
```diff
@@ -3703,10 +3703,10 @@ template <class OBJECT, class LAYOUT> void SetLayout(OBJECT *pObject, VkImage im
 }
 
 void SetLayout(const layer_data *dev_data, GLOBAL_CB_NODE *pCB, VkImageView imageView, const VkImageLayout &layout) {
-    auto iv_data = getImageViewData(dev_data, imageView);
-    assert(iv_data);
-    const VkImage &image = iv_data->image;
-    const VkImageSubresourceRange &subRange = iv_data->subresourceRange;
+    auto view_state = getImageViewState(dev_data, imageView);
+    assert(view_state);
+    const VkImage &image = view_state->create_info.image;
```
Comment (tobine) [inline layers/core_validation.cpp]: done

File: layers/core_validation.cpp [hunk 9]
```diff
@@ -9115,18 +9115,19 @@ static bool ValidateFramebufferCreateInfo(layer_data *dev_data, const VkFramebuf
             // attachmentCounts match, so make sure corresponding attachment details line up
             const VkImageView *image_views = pCreateInfo->pAttachments;
             for (uint32_t i = 0; i < pCreateInfo->attachmentCount; ++i) {
-                VkImageViewCreateInfo *ivci = getImageViewData(dev_data, image_views[i]);
-                if (ivci->format != rpci->pAttachments[i].format) {
+                auto view_state = getImageViewState(dev_data, image_views[i]);
+                auto ivci = view_state->create_info;
```
Comment (tobine) [inline layers/core_validation.cpp]: done

File: tests/layer_validation_tests.cpp (hunk 3 above)
Comment (tobine) [inline tests/layer_validation_tests.cpp]: gone

File: tests/layer_validation_tests.cpp (hunk 4 above)
Comment (tobine) [inline tests/layer_validation_tests.cpp]: done

**Pull request merged**